sentinel.go

```diff
@@ -289,19 +305,75 @@ func (c *sentinelClient) _refresh() (err error) {
 	return err
 }
 
-func (c *sentinelClient) listWatch(cc conn) (master string, sentinels []string, err error) {
+// listWatch will use sentinel to list current master|replica address along with sentinels address
+func (c *sentinelClient) listWatch(cc conn) (serverAddress string, sentinels []string, err error) {
 	ctx := context.Background()
 	sentinelsCMD := c.cmd.SentinelSentinels().Master(c.mOpt.Sentinel.MasterSet).Build()
 	getMasterCMD := c.cmd.SentinelGetMasterAddrByName().Master(c.mOpt.Sentinel.MasterSet).Build()
+	replicasCMD := c.cmd.SentinelReplicas().Master(c.mOpt.Sentinel.MasterSet).Build()
+
 	defer func() {
 		if err == nil { // not recycle cmds if error, since cmds may be used later in pipe. consider recycle them by pipe
 			cmds.PutCompleted(sentinelsCMD)
 			cmds.PutCompleted(getMasterCMD)
+			cmds.PutCompleted(replicasCMD)
 		}
 	}()
 
-	go func(cc conn) {
+	go c.watchEvents(ctx, cc)
+
+	commands := []Completed{
+		sentinelsCMD,
+		getMasterCMD,
+	}
+
+	if c.readonly {
+		commands = append(commands, replicasCMD)
+	}
+
+	resp := cc.DoMulti(ctx, commands...)
+	others, err := resp[0].ToArray()
+	if err != nil {
+		return "", nil, err
+	}
+	for _, other := range others {
+		if m, err := other.AsStrMap(); err == nil {
+			sentinels = append(sentinels, fmt.Sprintf("%s:%s", m["ip"], m["port"]))
+		}
+	}
+
+	// we return random slave address instead of master
+	if c.readonly {
+		replicaResponses, err := resp[2].ToArray()
+		if err != nil {
+			return "", nil, err
+		}
+
+		// choose a replica randomly
+		randomReplicaIndex := rand.Intn(len(replicaResponses))
+		m, err := replicaResponses[randomReplicaIndex].AsStrMap()
+		if err != nil {
+			return "", nil, err
+		}
+
+		return fmt.Sprintf("%s:%s", m["ip"], m["port"]), sentinels, nil
+	}
+
+	// otherwise send master as addrress
+	m, err := resp[1].AsStrSlice()
+	if err != nil {
+		return "", nil, err
+	}
+	return fmt.Sprintf("%s:%s", m[0], m[1]), sentinels, nil
+}
+
+// watchEvents will help subscribing to channel ONLY ONCE.
+func (c *sentinelClient) watchEvents(ctx context.Context, cc conn) {
```

Comment:
This is intentional because in the previous implementation, every time a message is received, eventually it will cause another launch of `cc.Receive(...)` which then will lead to multiple operations for a single event.

Reply:
Hi @smf8,

Since you mentioned another issue here, could we deal with it first and in a separate PR?

I just looked back into the `refresh` and `listWatch`, and I think they would only cause duplicated `cc.Receive(...)` when `_switchMaster` failed and we can quickly fix the duplication by always closing the previous `sConn` in the `_refresh`.

Reply:
Sorry, we need to send UNSUBSCRIBE first before we close the previous `sConn` so that the previous `cc.Receive(...)` can exit without error.

Reply:
How about only sending an `UNSUBSCRIBE`  command before starting `cc.Receive` in the main goroutine, if there wasn't any previous subscription it will execute without doing anything. So there won't be any need to close the connection.

Reply:
Yes, good idea.

Reply:
#249